### PR TITLE
Move marketing year to invoice line

### PR DESCRIPTION
**Pull request by johnwatson484**

### CODE REVIEW

File: app/processing/siti-agri/schemas/invoice-line.js [hunk 1]
```diff
@@ -8,5 +8,6 @@ module.exports = Joi.object({
   description: Joi.string().regex(/^[A-Z]{1}\d{2}\s-\s.+$/).required(),
   value: Joi.number().required(),
   convergence: Joi.boolean().optional(),
-  deliveryBody: Joi.string().regex(/^[A-Z]{2}\d{2}$/).required()
+  deliveryBody: Joi.string().regex(/^[A-Z]{2}\d{2}$/).required(),
+  marketingYear: Joi.number().integer().greater(2015).less(2099).required()
```
Comment (Kirsten1822) [inline app/processing/siti-agri/schemas/invoice-line.js:12]: Do you need to add the .less() part here?

Comment (johnwatson484) [reply app/processing/siti-agri/schemas/invoice-line.js:12]: Just added it so it's consistent with what we're already validating on the header level.

It's an interesting point.  I feel we need something, so let's say we get a crazy value like `33223` we probably want to reject that and not just accept it.  But we don't want to lock it down so much that we have to keep changing the schema.

I guess less than 2099 was agreed on for a reason and I didn't have a strong case for an alternative

File: docs/asyncapi.yaml [hunk 2]
```diff
@@ -254,10 +255,11 @@ components:
           items:
             type: object
             required:
-              - accountCode
               - description
               - schemeCode
               - value
+              - deliveryBody
+              - marketingYear
```
Comment (Kirsten1822) [inline docs/asyncapi.yaml:262]: I think there's a few other required missing from the list - should they be added?

File: docs/asyncapi.yaml [hunk 3]
```diff
@@ -413,7 +375,7 @@ components:
     Timestamp:
       type: string
       description: Current datetime
-      example: "2022-01-01T00:00:00.000Z"
+      example: '2022-01-01T00:00:00.000Z'
```
Comment (Kirsten1822) [inline docs/asyncapi.yaml:378]: Does the asyncapi document not already take this as a string or is it iffy?

File: docs/asyncapi.yaml [hunk 4]
```diff
@@ -425,36 +387,9 @@ components:
       type: number
       format: decimal
       description: Decimal net value of request before enrichment
-      minimum: 0.00
+      minimum: 0
```
Comment (Kirsten1822) [inline docs/asyncapi.yaml:390]: Why remove the decimals here?

File: docs/asyncapi.yaml [hunk 5]
```diff
@@ -469,10 +404,10 @@ components:
       type: string
       format: DD-MM-YYYY
       description: Date SITI Agri exported payment file
-      example: "2022-06-28"
+      example: '2022-06-28'
     Checkpoint:
       type: string
-      description: Service which raised the event, differs per environment
+      description: 'Service which raised the event, differs per environment'
```
Comment (Kirsten1822) [inline docs/asyncapi.yaml:410]: I don't think the single quotes are needed here

File: test/unit/processing/siti-agri/build-invoice-lines.test.js
Comment (Kirsten1822) [inline test/unit/processing/siti-agri/build-invoice-lines.test.js:1]: Does the let need to be inside the describe?

File: docs/asyncapi.yaml (hunk 3 above)
Comment (johnwatson484) [reply docs/asyncapi.yaml:378]: I didn't actually change any of these, I just removed the duplicate values, I guess some had quotes and some didn't.

I'll fix.

File: test/unit/processing/siti-agri/build-invoice-lines.test.js
Comment (Kirsten1822) [inline test/unit/processing/siti-agri/build-invoice-lines.test.js:1]: Could the example data used in invoiceLines be put in its own separate mock folder and referenced in? Or is that out of scope?

File: docs/asyncapi.yaml (hunk 4 above)
Comment (johnwatson484) [reply docs/asyncapi.yaml]: I didn't.  I removed the duplicate, but guess it didn't have decimals.  Added them

File: docs/asyncapi.yaml (hunk 5 above)
Comment (johnwatson484) [reply docs/asyncapi.yaml]: Agreed, another one I didn't change (I promise!).  Removed quotes

File: test/unit/processing/siti-agri/build-invoice-lines.test.js
Comment (johnwatson484) [reply test/unit/processing/siti-agri/build-invoice-lines.test.js:1]: I think it should.  I did think about doing it, but I didn't want to turn this into another mega review and go off re-writing all the tests.

I think I'll create an improvement ticket to specifically make more use of mocks in this repo's tests.

File: app/processing/siti-agri/schemas/invoice-line.js (hunk 1 above)
Comment (Kirsten1822) [reply app/processing/siti-agri/schemas/invoice-line.js:12]: For consistency sake I'd say leave it as is. The less than value doesn't matter so much, so long as it's not too constrictive like you said

File: docs/asyncapi.yaml (hunk 2 above)
Comment (Kirsten1822) [reply docs/asyncapi.yaml:262]: Fine by me :)

File: test/unit/processing/siti-agri/build-invoice-lines.test.js
Comment (Kirsten1822) [reply test/unit/processing/siti-agri/build-invoice-lines.test.js:1]: Sounds good to me

File: docs/asyncapi.yaml [hunk 6]
```diff
@@ -469,7 +404,7 @@ components:
       type: string
       format: DD-MM-YYYY
       description: Date SITI Agri exported payment file
-      example: "2022-06-28"
+      example: '2022-06-28'
```
Comment (Kirsten1822) [inline docs/asyncapi.yaml:407]: Do dates need the single quotes for it to appear properly in the example message?

File: docs/asyncapi.yaml [hunk 7]
```diff
@@ -513,9 +448,9 @@ components:
       example: G00 - Gross value of claim
     DueDate:
       type: string
-      format: DD/MM/YYYY, YYYY/MM/DD, DD-MM-YYYY or YYYY-MM-DD
+      format: 'DD/MM/YYYY, YYYY/MM/DD, DD-MM-YYYY or YYYY-MM-DD'
       description: Date request should be issued from
-      example: "2022-12-01"
+      example: '2022-12-01'
```
Comment (Kirsten1822) [inline docs/asyncapi.yaml:453]: See my comment here: https://github.com/DEFRA/ffc-pay-batch-processor/pull/84/files#r1211487299

File: docs/asyncapi.yaml [hunk 8]
```diff
@@ -513,9 +448,9 @@ components:
       example: G00 - Gross value of claim
     DueDate:
       type: string
-      format: DD/MM/YYYY, YYYY/MM/DD, DD-MM-YYYY or YYYY-MM-DD
+      format: 'DD/MM/YYYY, YYYY/MM/DD, DD-MM-YYYY or YYYY-MM-DD'
```
Comment (Kirsten1822) [inline docs/asyncapi.yaml:451]: Does this need the single quotes?

File: docs/asyncapi.yaml (hunk 6 above)
Comment (johnwatson484) [reply docs/asyncapi.yaml:407]: Seems not.   Changed

File: docs/asyncapi.yaml (hunk 7 above)
Comment (johnwatson484) [reply docs/asyncapi.yaml:453]: Nope, removed

File: docs/asyncapi.yaml (hunk 8 above)
Comment (johnwatson484) [reply docs/asyncapi.yaml:451]: It does not and this is also incorrect too.  Corrected.

**Review by Kirsten1822 — APPROVED** [review]
Happy to approve :)

**Pull request merged**